File: src/main/java/com/cstestforge/ado/service/AdoService.java
package com.cstestforge.ado.service;

import com.cstestforge.ado.model.*;

import java.util.List;

/**
 * Service for Azure DevOps integration
 */
public interface AdoService {

    /**
     * Get all ADO connections
     * 
     * @return List of ADO connections
     */
    List<AdoConnection> getConnections();
    
    /**
     * Get ADO connection by ID
     * 
     * @param id Connection ID
     * @return The connection or null if not found
     */
    AdoConnection getConnectionById(String id);
    
    /**
     * Create a new ADO connection
     * 
     * @param connection Connection to create
     * @return Created connection with ID
     */
    AdoConnection createConnection(AdoConnection connection);
    
    /**
     * Update an existing ADO connection
     * 
     * @param id Connection ID
     * @param connection Updated connection data
     * @return Updated connection
     */
    AdoConnection updateConnection(String id, AdoConnection connection);
    
    /**
     * Delete an ADO connection
     * 
     * @param id Connection ID
     * @return True if deleted successfully
     */
    boolean deleteConnection(String id);
    
    /**
     * Validate ADO connection credentials
     * 
     * @param url Azure DevOps URL
     * @param pat Personal access token
     * @param organizationName Organization name
     * @param projectName Project name
     * @return True if connection is valid
     */
    boolean validateConnection(String url, String pat, String organizationName, String projectName);
    
    /**
     * Get ADO projects for a connection
     * 
     * @param connectionId Connection ID
     * @return List of ADO projects
     */
    List<AdoProject> getProjects(String connectionId);
    
    /**
     * Get ADO test plans for a project
     * 
     * @param connectionId Connection ID
     * @param projectId Project ID
     * @return List of test plans
     */
    List<AdoTestPlan> getTestPlans(String connectionId, String projectId);
    
    /**
     * Get ADO test suites for a test plan
     * 
     * @param connectionId Connection ID
     * @param projectId Project ID
     * @param testPlanId Test plan ID
     * @return List of test suites
     */
    List<AdoTestSuite> getTestSuites(String connectionId, String projectId, String testPlanId);
    
    /**
     * Get ADO pipelines for a project
     * 
     * @param connectionId Connection ID
     * @param projectId Project ID
     * @return List of pipelines
     */
    List<AdoPipeline> getPipelines(String connectionId, String projectId);
    
    /**
     * Get synchronization configuration for a project
     * 
     * @param projectId Project ID
     * @return Sync configuration or null if not configured
     */
    SyncConfig getSyncConfig(String projectId);
    
    /**
     * Save synchronization configuration for a project
     * 
     * @param projectId Project ID
     * @param config Sync configuration
     * @return Updated sync configuration
     */
    SyncConfig saveSyncConfig(String projectId, SyncConfig config);
    
    /**
     * Get synchronization status for a project
     * 
     * @param projectId Project ID
     * @return Sync status or null if not available
     */
    SyncStatus getSyncStatus(String projectId);
    
    /**
     * Start synchronization for a project
     * 
     * @param projectId Project ID
     * @return True if sync was started successfully
     */
    boolean startSync(String projectId);
    
    /**
     * Get pipeline configuration for a project
     * 
     * @param projectId Project ID
     * @return Pipeline configuration or null if not configured
     */
    PipelineConfig getPipelineConfig(String projectId);
    
    /**
     * Save pipeline configuration for a project
     * 
     * @param projectId Project ID
     * @param config Pipeline configuration
     * @return Updated pipeline configuration
     */
    PipelineConfig savePipelineConfig(String projectId, PipelineConfig config);
    
    /**
     * Trigger a pipeline run for a project
     * 
     * @param projectId Project ID
     * @return True if pipeline was triggered successfully
     */
    boolean triggerPipeline(String projectId);
} 

File: frontend/src/services/adoService.ts
import api from './api';
import {
  AdoConnection,
  AdoProject,
  AdoTestPlan,
  AdoTestSuite,
  AdoPipeline,
  SyncConfig,
  SyncStatus,
  PipelineConfig
} from '../context/AdoIntegrationContext';

const ADO_API_BASE = '/api/ado';

/**
 * Service for Azure DevOps integration
 */
const adoService = {
  /**
   * Get all ADO connections
   */
  async getConnections(): Promise<AdoConnection[]> {
    const response = await api.get(`${ADO_API_BASE}/connections`);
    return response.data;
  },

  /**
   * Create a new ADO connection
   */
  async createConnection(connection: Omit<AdoConnection, 'id' | 'createdAt' | 'updatedAt'>): Promise<AdoConnection> {
    const response = await api.post(`${ADO_API_BASE}/connections`, connection);
    return response.data;
  },

  /**
   * Update an existing ADO connection
   */
  async updateConnection(id: string, connection: Partial<AdoConnection>): Promise<AdoConnection> {
    const response = await api.put(`${ADO_API_BASE}/connections/${id}`, connection);
    return response.data;
  },

  /**
   * Delete an ADO connection
   */
  async deleteConnection(id: string): Promise<boolean> {
    const response = await api.delete(`${ADO_API_BASE}/connections/${id}`);
    return response.status === 200;
  },

  /**
   * Validate an ADO connection
   */
  async validateConnection(url: string, pat: string, organizationName: string, projectName: string): Promise<boolean> {
    const response = await api.post(`${ADO_API_BASE}/connections/validate`, {
      url,
      pat,
      organizationName,
      projectName
    });
    return response.data.valid;
  },

  /**
   * Get projects for a connection
   */
  async getProjects(connectionId: string): Promise<AdoProject[]> {
    const response = await api.get(`${ADO_API_BASE}/connections/${connectionId}/projects`);
    return response.data;
  },

  /**
   * Get test plans for a project
   */
  async getTestPlans(connectionId: string, projectId: string): Promise<AdoTestPlan[]> {
    const response = await api.get(`${ADO_API_BASE}/connections/${connectionId}/projects/${projectId}/test-plans`);
    return response.data;
  },

  /**
   * Get test suites for a test plan
   */
  async getTestSuites(connectionId: string, projectId: string, testPlanId: string): Promise<AdoTestSuite[]> {
    const response = await api.get(
      `${ADO_API_BASE}/connections/${connectionId}/projects/${projectId}/test-plans/${testPlanId}/test-suites`
    );
    return response.data;
  },

  /**
   * Get pipelines for a project
   */
  async getPipelines(connectionId: string, projectId: string): Promise<AdoPipeline[]> {
    const response = await api.get(`${ADO_API_BASE}/connections/${connectionId}/projects/${projectId}/pipelines`);
    return response.data;
  },

  /**
   * Get synchronization configuration for a project
   */
  async getSyncConfig(projectId: string): Promise<SyncConfig> {
    const response = await api.get(`${ADO_API_BASE}/sync/config/${projectId}`);
    return response.data;
  },

  /**
   * Save synchronization configuration for a project
   */
  async saveSyncConfig(projectId: string, config: SyncConfig): Promise<SyncConfig> {
    const response = await api.post(`${ADO_API_BASE}/sync/config/${projectId}`, config);
    return response.data;
  },

  /**
   * Get synchronization status for a project
   */
  async getSyncStatus(projectId: string): Promise<SyncStatus> {
    const response = await api.get(`${ADO_API_BASE}/sync/status/${projectId}`);
    return response.data;
  },

  /**
   * Start synchronization for a project
   */
  async startSync(projectId: string): Promise<boolean> {
    const response = await api.post(`${ADO_API_BASE}/sync/start/${projectId}`);
    return response.status === 200;
  },

  /**
   * Get pipeline configuration for a project
   */
  async getPipelineConfig(projectId: string): Promise<PipelineConfig> {
    const response = await api.get(`${ADO_API_BASE}/pipeline/config/${projectId}`);
    return response.data;
  },

  /**
   * Save pipeline configuration for a project
   */
  async savePipelineConfig(projectId: string, config: PipelineConfig): Promise<PipelineConfig> {
    const response = await api.post(`${ADO_API_BASE}/pipeline/config/${projectId}`, config);
    return response.data;
  },

  /**
   * Trigger a pipeline run for a project
   */
  async triggerPipeline(projectId: string): Promise<boolean> {
    const response = await api.post(`${ADO_API_BASE}/pipeline/trigger/${projectId}`);
    return response.status === 200;
  }
};

export default adoService; 